lib/network/modules/layout-engine/index.ts

```diff
@@ -73,19 +84,31 @@ export function fillLevelsByDirectionLeaves(nodes: Map<Id, Node>): Levels {
 
 /**
  * Assign levels to nodes according to their positions in the hierarchy. Roots will be lined up at the top and all nodes as close to their parents as possible.
- *
  * @param nodes - Visible nodes of the graph.
  * @returns Populated node levels.
  */
 export function fillLevelsByDirectionRoots(nodes: Map<Id, Node>): Levels {
+  // Collect valid root nodes
+  const entryNodes: Node[] = [];
+  nodes.forEach((node) => {
+    // Pick only leaves (nodes without children).
+    const isRoot = node.edges
+      // Take only visible nodes into account.
+      .filter((edge) => nodes.has(edge.toId))
+      // Check that all edges lead from this node (root).
+      .every((edge) => edge.from === node);
+    if (isRoot) {
+      entryNodes.push(node);
+    }
+  });
+
+  if (entryNodes.length === 0) {
+    return fillLevelsByDirectionCyclic(nodes, {});
+  }
+
   return fillLevelsByDirection(
     // Pick only roots (nodes without parents).
-    (node): boolean =>
-      node.edges
-        // Take only visible nodes into account.
-        .filter((edge): boolean => nodes.has(edge.toId))
-        // Check that all edges lead from this node (root).
-        .every((edge): boolean => edge.from === node),
+    (node): boolean => entryNodes.includes(node),
```

Comment:
What is the purpose of this? Won't the condition at the end catch cyclic graphs every time this would? It seems to me like just added overhead for people who are not abusing this layout for cyclic graphs. Off course I may be missing something.